clients/unshaded/src/main/java/tachyon/client/BlockOutStream.java

```diff
@@ -16,239 +16,66 @@
-    mLocalFileChannel = mCloser.register(mLocalFile.getChannel());
-    // change the permission of the temporary file in order that the worker can move it.
-    CommonUtils.changeLocalFileToFullPermission(mLocalFilePath);
-    // use the sticky bit, only the client and the worker can write to the block
-    CommonUtils.setLocalFileStickyBit(mLocalFilePath);
-    LOG.info(mLocalFilePath + " was created!");
-    mAvailableBytes += initialBytes;
-
-    long allocateBytes = mTachyonConf.getBytes(Constants.USER_FILE_BUFFER_BYTES, Constants.MB) + 4L;
-    mBuffer = ByteBuffer.allocate(Ints.checkedCast(allocateBytes));
-  }
-
-  private synchronized void appendCurrentBuffer(byte[] buf, int offset, int length)
-      throws IOException {
-    if (mAvailableBytes < length) {
-      long bytesRequested = mTachyonFS.requestSpace(mBlockId, length - mAvailableBytes);
-      if (bytesRequested + mAvailableBytes >= length) {
-        mAvailableBytes += bytesRequested;
-      } else {
-        mCanWrite = false;
-        throw new IOException(String.format("No enough space on local worker: fileId(%d)"
-            + " blockId(%d) requestSize(%d)", mFile.mFileId, mBlockId, length - mAvailableBytes));
-      }
-    }
-
-    MappedByteBuffer out = mLocalFileChannel.map(MapMode.READ_WRITE, mInFileBytes, length);
-    out.put(buf, offset, length);
-    CommonUtils.cleanDirectBuffer(out);
-    mInFileBytes += length;
-    mAvailableBytes -= length;
-  }
-
-  @Override
-  public void cancel() throws IOException {
-    if (!mClosed) {
-      mCloser.close();
-      mClosed = true;
-      mTachyonFS.cancelBlock(mBlockId);
-      LOG.info(String.format("Canceled output of block. blockId(%d) path(%s)", mBlockId,
-          mLocalFilePath));
-    }
+    return get(tachyonFile, opType, blockIndex,
+        tachyonConf.getBytes(Constants.USER_QUOTA_UNIT_BYTES, 8 * Constants.MB), tachyonConf);
   }
 
   /**
-   * @return true if the stream can write and is not closed, otherwise false
+   *
+   * @param tachyonFile The file this block belongs to.
+   * @param opType The type of write.
+   * @param blockIndex The index of the block in the tachyonFile.
+   * @param initialBytes The initial size (in bytes) that will be allocated to the block.
+   * @param tachyonConf The TachyonConf instance.
+   * @return A new {@link LocalBlockOutStream} or {@link RemoteBlockOutStream}.
+   * @throws IOException
    */
-  public boolean canWrite() {
-    return !mClosed && mCanWrite;
-  }
-
-  @Override
-  public void close() throws IOException {
-    if (!mClosed) {
-      if (mBuffer.position() > 0) {
-        appendCurrentBuffer(mBuffer.array(), 0, mBuffer.position());
-      }
-      mCloser.close();
-      mTachyonFS.cacheBlock(mBlockId);
-      mClosed = true;
+  public static BlockOutStream get(TachyonFile tachyonFile, WriteType opType, int blockIndex,
+      long initialBytes, TachyonConf tachyonConf) throws IOException {
+
+    if (tachyonFile.mTachyonFS.hasLocalWorker()
```

Comment:
Yeah, that could be a useful feature, but I don't think it can be done simply right now.

In the future, maybe we can make the client re-connect to a different worker if the local worker does not have space. This may have to touch higher levels of the code, like the TFS, and/or master.